pkg/controllermanager/controller/managedseedset/managedseedset_predicates.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseedset
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	operationshoot "github.com/gardener/gardener/pkg/operation/shoot"
+)
+
+func filterShoot(obj, controller client.Object, deleted bool) bool {
+	shoot, ok := obj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return false
+	}
+	set, ok := controller.(*seedmanagementv1alpha1.ManagedSeedSet)
+	if !ok {
+		return false
+	}
+
+	if set.Status.PendingReplica == nil || set.Status.PendingReplica.Name != shoot.Name {
```

Comment:
From the perspective of reconciling scaling in or out (and I suspect updates also, but I can't be sure before I implement it), we should not enqueue such a shoot. Let's consider the different cases:

* If there are no pending replicas, the scaling in / out has finished, and until the next scaling in / out there is no reason for the set to be interested in any of the shoot's health.
* If there is a different pending replica with a higher ordinal, the replica with changing health has already been healthy during this scaling in / out, so no reason to be interested in it any more.
* If there is a different pending replica with a lower ordinal, rescheduling the set would lead to a wasted reconcile, since it would most likely simply reach this lower ordinal replica and stop there. The replica with changing health would not be reached.

Regarding the `readyReplicas` field - yes it may be "out of date" during periods of time while some shoots are unhealthy. Is this really so important?

Reply:
Hmm, ok. My understanding of the `readyReplicas` field was, that it would also take the shoots' health status into account. Meaning, it wouldn't just simply show how many replicas it has successfully created (or rolled out) but rather how many of them are actually currently ready (in the sense of healthy).
But it seems you had something else in mind (basically what was my understanding of the `currentReplicas` field).

The question is, what the meaning of the status field should be now. (Can discuss this in our sync tomorrow).

Reply:
Yes, let's discuss the meaning of the field in details. In general, I am rather inclined to avoid triggering additional reconciles just for this one field, but I am not 100% sure.

Reply:
We agreed to change this as suggested, and also to change the actuator so that it only acts on the pending replica, if there is one, instead of iterating over all replicas.

Reply:
I realized we missed something rather important so far in this discussion. What happens if I e.g. delete one of the managed seeds manually (or change it, or change its shoot)?

I think the expectation is that the `ManagedSeedSet` controller will eventually restore the set to the state in the resource, including recreating manually deleted objects. This cannot happen if the controller only acts on the pending replica, if there is one. Therefore, IMO the controller must, as originally implemented, iterate over all replicas to find the first one that it needs to act upon. In addition, the filter functions should be extended to make sure such events enqueue the set.

For now I will implement it like this and test it a bit, we should discuss this further.

Reply:
I added a new commit that improves upon the above by ensuring that updates to non-pending replicas don't interrupt scaleouts / scaleins, while at the same time making sure that any non-ready non-pending replicas are eventually taken care of. Please take a look.